# Python Referenzen

Dieses Notebook wiederholt Grundkonzepte der Programmiersprache Python und deren Syntax (Schreibweise).
Obwohl es nicht eine vollständige Zusammenfassung des bereits erlenten Stoffs ist, soll dieses Notebook als Hilfe für zukünftige Programmieraufgaben dienen. 

## Inhaltsverzeichnis

- [1. Einleitung](#einleitung)
  - [Python](#python)
- [2. Erste Schritte](#first-steps)
  - [print() und input()](#print-input)
- [3. Syntax](#syntax)
  - [Variablen und Operatoren](#var-und-operatoren)
- [4. Datentypen](#datentypen)
  - [Zahlen](#zahlen)
  - [Zeichenketten](#zeichenketten)
  - [Listen](#listen)
  - [Wahrheitswerte](#booleans)
- [5. Programmierkurs](#programmierkurs)
  - [Verzweigungen](#verzweigungen)
  - [Schleifen](#schleifen)
    - [for-Schleife](#for-schleife)
    - [while-Schleife](#while-schleife)
- [6. Weiterführende Programmierung](#weiterfuehrende-programmierung)
  - [Funktionen](#funktionen)
    - [Parameter](#parameter)
    - [Rückgabewerte](#rueckgabewerte)
  - [Eingebaute Funktionen](#eingebaute-funktionen)
  - [Ausgabe und Formatierung](#ausgabe-formatierung)
    - [Funktion print()](#print)
  - [Module](#module)

# 1. Einleitung <a class="anchor" id="einleitung"></a>

**Hintergrund Programmiersprachen**<br>
Alle Berechnungen und Abläufe die auf einem Computer durchgeführt werden, werden vom sogenannten **Prozessor** ausgeführt. Dieser **Prozessor** (*CPU für **C**entral **P**rocessing **U**nit*) folgt Anweisungen und Befehlen die ihm zugewiesen werden. Der **Prozessor** versteht Instruktionen aber nur, wenn diese in **Bit-Wörtern** (001101011..) geschrieben sind. Solche **Bit-Wörter** könn(t)en mit der [**Maschinensprache**](https://de.wikipedia.org/wiki/Maschinensprache) vom Programmierer geschrieben werden. Diese **Maschinensprache** ist jedoch sehr schwierig zu erlernen und braucht häufig mehrere Zeilen um eine einfache Berechnung, wie die Addition von zwei Zahlen, auf dem Prozessor durchzuführen.
Deshalb wurden *höhere Programmiersprachen* entwickelt.<br><br> Höhere Programmiersprachen, oder einfach *Programmiersprachen*, sind **formale Sprachen** zur Formulierung von Datenstrukturen und Algorithmen. Sie haben das Ziel, dem Menschen eine *leichter verständliche Ausdrucksweise* bei der Programmierung zu erlauben. Nachdem der Entwickler sein Programm, in der *höheren Programmiersprache*, fertig geschrieben hat, muss es noch in *Bit-Wörter* **übersetzt** werden, damit es vom **Prozessor** auch wirklich ausgeführt werden kann. Die Übersetzung wird von einem sogenannten **Interpreter** oder **Compiler** durchgeführt.

## 1.1 Python <a class="anchor" id="python"></a>

[Python](https://www.python.org) ist eine sehr einfach zu erlernende Programmiersprache und für den Einstieg in die Welt der Pragrammierung ideal geeignet. Trotz ihrer Einfachheit bietet diese Sprache auch die Möglichkeit, komplexe Programme für vielfältige Anwendungsgebiete zu schreiben.<br>

Um Python Programme auszuführen, muss auf dem Rechner ein [**Python-Interpreter**](https://www.python.org/downloads/) (*Übersetzer*) installiert werden. Dieser wird unsere Programme in *Maschinensprache (Bit-Wörter)* übersetzen, sodass unsere Anweisungen und Befehle auf dem Rechner ausgeführt werden können.

<img src="interpreter.PNG"/>

Wichtige Vorteile von Python:
- Python verügt über eine einfache und eindeutige [Syntax](#syntax): <br> Die Schreibweise von Python beschränkt sich auf einfache, klare Anweisungen und häufig auf einen einzigen möglichen Lösungsweg. Dieser prägt sich schnell ein und wird dem Entwickler vertraut.
- Klare Strukturen: Python verlangt vom Entwickler, in einer gut lesbaren Struktur zu schreiben. Die Anordnung der Programmzeilen ergibt zugleich die logische Struktur des Programms.

Python Programme können in einem Beliebigen Text-Editor geschrieben werden und haben die Dateiendung `.py`. Für die Programmierung wurden spezifische Text-Editoren entwickelt, die dem Entwickler die Arbeit leichter machen sollen. Ein solcher Text-Editor nennen wir **Entwicklungsumgebung**. In diesem Kurs verwenden wir die Entwicklungsumgebung [Visual Studio Code](https://code.visualstudio.com).

In diesem Kurs schreiben wir oft nicht einzelne Python Programme (einzelne Dateien mit der Dateiendung `.py`), wir arbeiten stattdessen mit *Jupyter Notebooks* (Dateiendung `.ipynb`). Diese ermöglichen uns die Vereinigung von erzählendem Text und ausführbaren Programmier-Zellen. Zudem können wir einzelne **Code-Blöcke** individuell ausführen lassen. Weitere Informationen zu *Jupyter Notebooks* findest du im Notebook *Jupyter Notebooks*. 

# 2. Erste Schritte <a class="anchor" id="first-steps"></a>

Bevor wir uns mit der Syntax und den Datentypen von Python vertraut machen, befassen wir uns mit der `input()` und der `print()` Funktion.
Diese beiden Funktionen ermöglichen **Ein- und Ausgabe-Funktionalitäten**, sodass ein Programm mit der *Aussenwelt* kommunizieren kann.<br>
Beide Funktionen sind fundamentale Bestandteile der Programmiersprache, d.h. der Python **Interpreter** versteht die Befehle `input()` und `print()`.<br>
`print()` und `input()` sind sogenannte [**eingebaute Funktionen**](#eingebaute-funktionen).

## 2.1 print() und input() <a class="anchor" id="print-input"></a>

### Ausgabe mit print()

In Python wird eine **Ausgabe** mit der Funktion `print()` erzeugt. Sie schreibt den Ausdruck in den einfachen Klammern `()` in die Kommandozeile.
Der Ausdruck kann ein **beliebiger [Datentyp](#datentypen)** sein.

In [47]:
# Eine Zeichenkette muss in einfachen oder doppelten Anführungszeichen geschrieben werden
print("Hallo Welt")

Hallo Welt


Wenn wir einen Python-Ausdruck (Berechnung oder Variable) an die `print()` Funktion übergeben, wird der Ausdruck zuerst evaliert (berechnet), und der resultierende Wert wird anschliessend in die Kommandozeile geschrieben.

In [48]:
# Wir können auch Python Anweisungen als Ausdruck an die 'print'-Funktion übergeben.
print(5 + 8)

13


Die Funktion kann auch mehrere Ausdrücke hintereinander in die Kommandozeile schreiben. Mehere Ausdrücke können mit einem Komma getrennt werden.
Wenn mehere Ausdrücke mit dem Komma getrennt werden, wird automatisch ein Leerzeichen zwischen den Ausdrücken eingefügt.

In [49]:
print("Hallo", "Welt", 15)

Hallo Welt 15


Im Verlauf von diesem Notebook wirst du noch weitere Besipiele der `print()` Funktion sehen.

### Eingabe mit input()

In Python wird eine **Eingabe** mit der Funktion `input()` erzeugt. Sobald die Funktion ausgeführt wird, wartet das Programm, bis der Benutzer etwas eingegeben hat. Der resultierende Wert der `input()`-Funktion ist eine [Zeichenkette](#zeichenketten).

In [50]:
# Wir speichern die Eingabe vom Benutzer in der Variable mit dem Namen eingegebenerWert
eingegebenerWert = input()
print(eingegebenerWert)

text


Wenn du für die Eingabe vom Benutzer auch noch eine Anweisung geben willst, kannst du dies als Parameter an der Funktion `input()` übergeben. Parameter schreiben wir wie bei der `print()` Funktion in die einfachen Klammern `()`.

In [51]:
# Wir speichern die Eingabe vom Benutzer in der Variable mit dem Namen name
name = input("Wie lautet dein Name?")
print(name)

Mauro


Im Verlauf dieses Noteboooks wirst du weitere Beispiele der Funktion `input()` sehen. Fürs erste reicht es, wenn du verstehst, wie du in einem Python Programm eine **Eingabe** und eine **Ausgabe** erzeugen kannst.

# 3. Syntax <a class="anchor" id="syntax"></a>

Als Syntax bezeichnen wir die *Schreibweise* von einer Programmiersprache. Die Syntax gibt uns vor, welche Ausdrücke wir in welchem Kontext verwenden können. Wir müssen die Syntax strikt befolgen, wenn wir wollen, dass der Python-**Interpreter** unser Programm verstehen kann.

## 3.1 Variablen und Operatoren <a class="anchor" id="var-und-operatoren"></a>

Wenn wir gewisse Werte mehrere Male in unserem Programm verwenden wollen, ergibt es Sinn, wenn wir diese Werte Speichern können. Dazu dienen **Variablen**. <br>
**Operatoren**, wie +, -, /, \*, dienen zur Ausführung von Berechnungen.

Die Namen von **Variablen** kannst du frei wählen. Es macht jedoch sinn, wenn du jeweils einen Namen wählst, der die Variable direkt beschreibt.

### Variablen zuweisen

Wenn wir einen Wert in einer Variable speichern wollen verwenden wir das Gleichheitszeichen. Auf der linken Seite des Gleichheitszeichens befindet sich der Name der Variable, und auf der rechten Seite der Wert, der in dieser Variable gespeichert wird.

In [52]:
# Eine einfache Zuweisung
# Der Name der Variable ist alter; 27 der Wert der in der Variable gespeichert wird
alter = 27

Im folgenden Code-Beispiel wird eine einfache Berechnung mithilfe eines **Operators** (*+ Zeichen*) durchgeführt. Das Ergebnis der Berechnung wird mit dem Gleichheitszeichen einer **Variablen** zugewiesen. Es erfolgt eine Ausgabe.

In [53]:
# Werte
a = 5
b = 3

# Berechnung
c = a + b

# Ausgabe
print("Die Aufgabe:", a, "+", b)
print("Das Ergebnis:", c)

Die Aufgabe: 5 + 3
Das Ergebnis: 8


*Hinweis*: Wie du bereits erkennen kannst, habe ich hier den *Komma*-Seperator für die **Ausgabe** verwendet. Das habe ich so gemacht, weil ich wollte, dass der Wert von a resp. b als solche in die Kommandozeile geschrieben werden. Überlege dir an dieser Stelle, warum im nächsten Beispiel `Die Aufgabe: 8` erscheint, und nicht `Die Aufgabe 5 + 3`. Wenn du den Unterschied nicht verstehst, solltest du dich melden. :)

In [54]:
print("Die Aufgabe:", a + b)

Die Aufgabe: 8


# 4. Datentypen <a class="anchor" id="datentypen"></a>

Beschäftigen wir uns noch einmal mit den Eigenschaften und Vorteilen der verschiedenen Datentypen. Es werden Operationen, Funktionen und Operatoren für die jeweiligen Datentypen vorgestellt.
Wenn du einen Wert in einer Variable speicherst, erkennt Python direkt, welcher Datentyp dabei verwendet werden soll.

*Übrigens*: Mit der Funktion `type()` kannst du jeweils überprüfen, welcher Datentyp in einer Variable gespeichert ist.

## 4.1 Zahlen <a class="anchor" id="zahlen"></a>

In diesem Abschnitt befassen wir uns kurz mit ganzen Zahlen und Zahlen mit Nachkommastellen. Es werden auch zwei **Operatoren** eingeführt, die mit Zahlen angewendet werden können.

### Ganze Zahlen

Als Datentyp für ganze Zahlen dient `int` (vom englisch *integer* für ganze Zahl).
*Integers* werden automatisch erstellt, wenn eine Ganze Zahl einer Variable zugeweisen wird.
Mit *ganzen Zahlen* können wir alle üblichen Operationen durchführen.

In [55]:
# In der Variable 'alter' wird eine ganze Zahle gespeichert
alter = 27
print(type(alter))

# In der Variable 'multiplikator' wird die Zahl 4 gespeichert
multiplikator = 4

# eine einfache Berechnung (Der Wert wird in der Variable 'resultat' gespeichert)
resultat = alter * multiplikator
print("Das Resultat der Multiplikation ist:", resultat)
print("Der Datentyp der Multiplikation ist:", type(resultat))

# Diese Division wird eine Zahl mit Nachkommastellen erzeugen, deshalb wird sie nicht mehr vom typ 'int' sein
resultat = alter / multiplikator
print("Das Resultat der Division ist:", resultat)
print("Der Datentyp der Division ist:", type(resultat))

<class 'int'>
Das Resultat der Multiplikation ist: 108
Der Datentyp der Multiplikation ist: <class 'int'>
Das Resultat der Division ist: 6.75
Der Datentyp der Division ist: <class 'float'>


#### Modulo operator

Wenn wir nun die Division $27 / 4$ durchführen erhalten wir $6.75$, also keine ganze Zahl. 
Anders ausgedrückt kann dieses Ergebnis auch $6 R 3$ lauten. Das heisst, bei der Division $27 / 4$ erhalten wir die ganze Zahl $6$ und den Rest $3$.
Der **Modulo Operator** (Syntax %) ermöglicht uns, den Wert des Rests zu erhalten. 

In [56]:
print("Modulo liefert:", alter % multiplikator)

Modulo liefert: 3


Überlege dir an dieser Stelle wie du mit dem **Modulo Operator** herausfinden kannst, ob eine Zahl ein Teiler von einer anderen Zahl ist. Gerne kannst du hier eine Code-Zelle einfügen, und deine Überlegungen ausprobieren.

*Hinweis*: Versuche zwei Variablen zu erstellen (zBsp. `a` und `b`), in denen du jeweils eine ganze Zahl speicherst. Versuche mit dem Modulo Operator herauszufinden, ob `b` ein Teiler von `a` ist.

In [57]:
# Code-Zelle 

### Zahlen mit Nachkommastellen

Der Datentyp für Zahlen mit Nachkommastellen heisst `float`. Diese sogenannten *Fliesskommazahlen* werden mithilfe des Dezimalpunkts geschrieben.
Genau wie bei den ganzen Zahlen können wir die bekannten Operatoren für diesen Datentyp verwenden.

In [58]:
kommazahl = 7.5
print(type(kommazahl))

<class 'float'>


#### \*\* Operator

Für alle Zahlen existiert der Potenz-Operator \*\*. Mit diesem Operator können wir Potenzen berechnen.

In [59]:
# Der Variablen z wird nacheinander das Ergebnis verschiedenen Exponentialrechnungen zugewiesen. 
# Anschliessend wird ihr jeweils aktueller Wert mit einem Kommentar ausgegeben
z = 5 ** 3
print("5 hoch 3 =", z)
z = 5.2 ** 3
print("5.2 hoch 3 =", z)
z = -5.2 ** 3
print("-5.2 hoch 3 =", z)
z = 5.2 ** 3.8
print("5.2 hoch 3.2 =", z)

5 hoch 3 = 125
5.2 hoch 3 = 140.608
-5.2 hoch 3 = -140.608
5.2 hoch 3.2 = 525.7904646699526


## 4.2 Zeichenketten <a class="anchor" id="zeichenketten"></a>

Zeichenkettern sind **Sequenzen** von einzelnen Zeichen - also Texte. Auch andere Datentypen gehören zu den **Sequenzen** (beispielsweise [Listen](#listen)). Anhand von Zeichenketten werden wir auch eine kurze Einführung in die **Sequenzen** machen.

Zeichenketten sind Objekte des Datentyps `str`. Strings bestehen aus mehreren Zeichen oder Wörtern. Sie werden gekennzeichnet, indem man sie in einfache, doppelte oder dreimal doppelte Anführungszeichen setzt.

In [60]:
t1 = "Hallo Welt"
t2 = 'Auch das ist eine Zeichenkette'
t3 = """Diese Zeichenkette
        steht in
        mehreren Zeilen"""
t4 = 'Hier sind "doppelte Hochkommata" gespeichert'

print("Bitte geben Sie einen Text ein")
t5 = input()

print("t1:", t1)
print("t2:", t2)
print("t3:", t3)
print("t4:", t4)
print("t5:", t5)

print("Typ:", type(t1))

Bitte geben Sie einen Text ein
t1: Hallo Welt
t2: Auch das ist eine Zeichenkette
t3: Diese Zeichenkette
        steht in
        mehreren Zeilen
t4: Hier sind "doppelte Hochkommata" gespeichert
t5: text
Typ: <class 'str'>


Die Zeichnkette `t4` verdeutlicht den Vorteil, den das Vorhandensein mehrerer Alternativen bietet: Die doppelten Hochkommata sind hier Bestandteil des Texts und werden auch ausgegeben.

### Operatoren <a class="anchor" id="operatoren-sequenzen"></a>

Die Operatoren $+$ und $*$ dienen zu Verkettung mehrerer **Sequenzen** bzw. zur Vervielfachnung einer **Sequenz**. Mithilfe des Operators `in` kannst du feststellen, ob ein bestimmtes Element in der Sequenz enthalten ist. Zudem kannst du mit dem Operator `not in` feststellen, ob das Element nicht in einer Sequenz enthalten ist.

**Achtung**: Der $+$ Operator funktioniert nur dann, wenn beide Elemente vom Typ `str` sind. Der $*$ Operator hingegebn multipliziert eine Zeichnkette mit einer Zahl.

In [61]:
# Operatoren + und *
t1 = "Teil 1"
t2 = "Teil 2"
tgesamt = t1 + ", " + t2

t3 = "-oooo-"
t4 = "***"
# 1-mal t4 + 3-mal t3 + 1-mal t4
tlinie = t4 + t3 * 3 + t4

print(tgesamt)
print(tlinie)

# Operator in
tname = "Robinson Crusoe"
print("Text:", tname)

if "b" in tname:
    print("b: ist enthalten")

if "p" not in tname:
    print("p: ist nicht enthalten")

Teil 1, Teil 2
***-oooo--oooo--oooo-***
Text: Robinson Crusoe
b: ist enthalten
p: ist nicht enthalten


### Operationen auf Sequenzen

#### Slices und Index

Sequenzen sind eigentlich eine Anweinanderkettung von Elementen. Das ermöglicht uns, einzelne Elemente in der Kette abzufragen. Alle Elemente innerhalb der Sequenz werden dabei **nummeriert**. Diese **Nummerierung** beginnt bei der Zahl $0$. Die **Nummer** von einem Element innerhalb der Kette bezeichnen wir als **Index**. Wenn wir ein einzelnes Element in der Zeichenkette abfragen wollen, müssen wir dies mit den eckigen Klammern (`[]`) machen. In die ekcigen Klammern schreiben wir jeweils den **Index** vom Element.

Ein Beispiel:

In [62]:
# Eine Zeichenkette wird gespeicher
tname = "Robinson Crusoe"

# Welches Element liegt an erster Stelle?
print("Erstes Element", tname[0])

# Das füfnfte Element hat den Index 4, da die Nummerierung bei 0 beginnt
print("Fünftes Element", tname[4])

Erstes Element R
Fünftes Element n


Teilbereiche von Sequenzen werden als *Slices* bezeichnet. Der Einsatz von Slices wird am Beispiel einer Zeichenkette verdeutlicht. Auf die gleiche Art und weise sind *Slices* auch auf andere Sequenz-Typen (wie Listen) anwendbar.

Das `slicing` funktioniert mit der gleichen Syntax wie die Abfrage von einem einzelnen Element (mit eckigen Klammern `[]`). Allerdings gilt: Wir müssen den Startwert und der Endwert des `slices` (Teilbereichs) definieren. Diese beiden Werte werden durch einen Doppelpunkt getrennt.

```python
zeichenkette[startwert:endwert]
```

Dabei gilt: 

**Startwert**: Erster Index der zum Slice gehört<br>
**Endwert**: Erster Index der **NICHT** mehr zum Slice gehört.

Also:

```python
zeichenkette[1:4]
```

wird uns alle Elemente an den Indizes 1, 2, 3 zurückgeben.

Aufgepasst: Wenn du **keinen** Startwert bzw. **keinen** Endwert definierst, werden **alle** Elemente nach dem Startwert (wenn kein Endwert angegeben ist), beziehungsweise **alle** Elemente vor dem Endwert (wenn kein Startwert angegeben ist) im *Slice* enthalten sein.

Ein Beispiel:

In [63]:
# Beispiel-Sequenz, hier Zeichenkette
tname = "Robinson Crusoe"
print("Text:", tname)

# Teilbereiche, Elemente
ts = tname[5:8]
print("[5:8]:", ts)
ts = tname[:8]
print("[:8]:", ts)
ts = tname[9:]
print("[9:]:", ts)
ts = tname[9]
print("[9]:", ts)

Text: Robinson Crusoe
[5:8]: son
[:8]: Robinson
[9:]: Crusoe
[9]: C


### Länge der Sequenz

Die eingebaute Funktion `len()` ermittelt die Anzahl der Elemente einer Sequenz. Im Fall einer Zeichenkette spricht man hierbei auch von der Länge der Zeichenkette.

**Syntax**: `len(sequenz)`

In [64]:
tname = "Robinson Crusoe"
print(len(tname))

15


### Umwandlung einer Zeichenkette in eine Zahl

Enthält eine Zeichenkette eine Zahl, etwa eine vom Benutzer eingegebene Zeichenkette, muss diese Zeichenkette zunächst Konvertiert werden.
Zur Umwandlung in eine ganze Zahl oder in eine Zahl mit Nachkommastellen dienen die beiden Funktionen `int()` bzw. `float()`. Anschliessend kann mit dieser Zahl gerechnet werden.


In [65]:
# Erste Zeichenkette
x = "15.3"

# Mit dieser Operation werden wir die Zeichnkette einfach Vervielfachen
ergebnis = x * 2
print(ergebnis)

# Wenn wir mit der Zahl rechnen wollen, muss Sie zuerst in einen 'float' Umgewandelt werden.
x = float(x)
ergebnis = x * 2
print(ergebnis)

# Zweite Zeichenkette
x = "17"

ergebnis = x * 2
print(ergebnis)

x = int(x)
ergebnis = x * 2
print(ergebnis)

15.315.3
30.6
1717
34


Stelle sicher, das du die einzelnen Code-Zellen verstehst. Zögere nicht und erstelle selber neue Code-Zellen wo du spielerisch die Funktionalitäten selber ausprobieren kannst.

## 4.3 Listen <a class="anchor" id="listen"></a>

Eine Liste ist eine **Sequenz** von Objekten in ihrere allgemeinsten Form. Sie kann Elemente von unterschiedlichen Datentypen enthalten. Eine Liste bietet vielfältige Möglichkeiten, u.a. die Funktionalitäten von ein- und mehrdimensionalen Feldern. Eine Liste kann an den eckigen Klammern `[]` erkennt werden. In diese Klammern können die Elemente der Liste Komma-getrennt angegeben werden.

Wenn du einzelne Elemente der Liste Abfragen willst musst du ebenfalls die eckigen Klammern `[]` verwenden. (Die Nummerierung beginnt auch hier bei 0!)

Beispiele:

In [66]:
# Liste von Zahlen
z = [3, 6, 12.5, -8, 5.5]
print(z)          # gesamte Liste
print(z[0])       # Das erste Element
print(z[2])       # Das dritte Element

# Liste von Zeichenketten
s = ["Bern", "Zürich", "Olten"]
print(s)

# Anzahl Elemente (auch hier verwenden wir die Funktion 'len()')
print("Anzahl:", len(s))

# Liste mit unterschiedlichen Datentypen
wilde_liste = [3, 7, "Hallo", "Welt", ["Liste", "in", "Liste"]]
print("Diese wilde Liste hat unterschiedliche Datentypen:", wilde_liste)

[3, 6, 12.5, -8, 5.5]
3
12.5
['Bern', 'Zürich', 'Olten']
Anzahl: 3
Diese wilde Liste hat unterschiedliche Datentypen: [3, 7, 'Hallo', 'Welt', ['Liste', 'in', 'Liste']]


Die Zeile `z[0]` gibt und den Wert vom ersten Element in der Liste zurück. Wie im Abschnitt [Operatoren](#operatoren-sequenzen) zu Sequenzen beschrieben, können die gleichen Operatoren auch für Listen benutzt werden.

**Neu bei Listen**:<br>
Im Gegensatz zu Zeichenketten können wir bei Listen auch einzelne Elemente **verändern**.
Wenn wir einzelne Elemente in der Liste verändern wollen, können wir das mit derselben Sytax (`[]`). 
Allerdings gilt: wenn wir einen neuen Wert einem Element zuweisen wollen, muss das Element auf der linken Seite des Gleichheitszeichens sein!

Hier ein Beispiel:

In [67]:
staedte = ["Bern", "Zürich", "Olten"]
print("Städte alt:", staedte)

# Der Wert des 2-ten Elements (Index = 1) abfragen und in einer Variable speichern
a_stadt = staedte[1]
print("A_Stadt:", a_stadt)

# Einen neue Wert am selben Ort abspeichern:
# Da wir den Wert eines Elements innerhalb der Liste verändern, muss staedte[1] auf der linken Seite erscheinen
staedte[1] = 'Lausanne'
print("Städte neu:", staedte)

Städte alt: ['Bern', 'Zürich', 'Olten']
A_Stadt: Zürich
Städte neu: ['Bern', 'Lausanne', 'Olten']


### Slices

Auch bei Listen können wir `Slices` (Teilbereiche) verwenden.

Beispiel:

In [68]:
staedte = ["Bern", "Zürich", "Olten", "Solothurn", "Genf", "Lausanne"]

# Wenn wir alle elemente vom 2-ten bis und mit zum 4-ten element abfragen wollen, müssen wir diese Werte als indizes angeben
print("Indizes 1, 2, 3:", staedte[1:4])

# Wenn wir keinen Startwert angeben, werden alle Element aufgezählt bis zum 'Endwert'!
print("Indizes 0, 1, 2:", staedte[:3])

# Analog können wir auch keinen Endwert angeben, dann gehören alle Indizes nach dm Startwert zum Slice
print("Indizes 3, 4, 5:", staedte[3:])

Indizes 1, 2, 3: ['Zürich', 'Olten', 'Solothurn']
Indizes 0, 1, 2: ['Bern', 'Zürich', 'Olten']
Indizes 3, 4, 5: ['Solothurn', 'Genf', 'Lausanne']


**Speziell bei Listen**: Wir können auch `slices` (Sublisten) innerhalb von Listen verändern.

In [69]:
# Originalliste
fr = ["Paris","Lyon","Marseille","Bordeaux"]
print("Original:")
print(fr)

# Ersetzen eines Teilbereiches durch eine Liste
fr[1:3] = ["Nancy","Metz","Gap"]
print("Teil ersetzt:")
print(fr)

Original:
['Paris', 'Lyon', 'Marseille', 'Bordeaux']
Teil ersetzt:
['Paris', 'Nancy', 'Metz', 'Gap', 'Bordeaux']


### Elemente löschen

Willst du Elemente oder Blöcke innerhalb einer Liste löschen kannst du das Schlüsselwort `del` (delete) verwenden.
Beispiel:

In [70]:
fr = ["Paris","Lyon","Marseille","Bordeaux"]
print("Original:", fr)

# Marseille wird gelöscht
del fr[2]
print("Element gelöscht:", fr)

# Ein Block wird gelöscht (Alle Elemente mit einem grösseren Index als 1)
del fr[1:]
print("Block gelöscht:", fr)

Original: ['Paris', 'Lyon', 'Marseille', 'Bordeaux']
Element gelöscht: ['Paris', 'Lyon', 'Bordeaux']
Block gelöscht: ['Paris']


### Funktionen für Listen

Listen beinhalten einige nützliche Funktionen. Wir wollen am folgenden Beispiel einige Funktionen kurz gemeinsam anschauen.

In [71]:
# Originallste
fr = ["Paris","Lyon","Marseille"]
print("Original:")
print(fr)

# Einsetzen eines Elements
fr.insert(1,"Nantes")
print("Nach Einsetzen:")
print(fr)

# Sortieren der Elemente
fr.sort()
print("Nach Sortieren:")
print(fr)

# Umdrehen der Liste
fr.reverse()
print("Nach Umdrehen:")
print(fr)

# Entfernen eines Elements
fr.remove("Nantes")
print("Nach Entfernen:")
print(fr)

# Ein Element am Ende hinzu
fr.append("Paris")
print("Ein Element hinzu:")
print(fr)

# Anzahl bestimmter Elemente
print("Anzahl Elemente Paris:", fr.count("Paris"))

# Suchen bestimmter Elemente
print("Erste Position Paris:", fr.index("Paris"))

Original:
['Paris', 'Lyon', 'Marseille']
Nach Einsetzen:
['Paris', 'Nantes', 'Lyon', 'Marseille']
Nach Sortieren:
['Lyon', 'Marseille', 'Nantes', 'Paris']
Nach Umdrehen:
['Paris', 'Nantes', 'Marseille', 'Lyon']
Nach Entfernen:
['Paris', 'Marseille', 'Lyon']
Ein Element hinzu:
['Paris', 'Marseille', 'Lyon', 'Paris']
Anzahl Elemente Paris: 2
Erste Position Paris: 0


- `insert(index, Element)`

Ein Element wird am angegebenen Index zur Liste hinzugefügt. Alle bisher in der Liste enthaltenen Element bleiben ebefalls in der Liste. Alle Elemente *hinter dem Index* werden nach dem `insert` um eine Stelle nach rechts verschoben.

- `sort()`

Die Liste wird sortiert. Falls es sich um eine Liste von Zeichenketten handelt, wird alphabetisch sortiert. Eine Liste von Zahlen wird nach Grösse sortiert.

- `reverse()`

Die Liste wird umgedreht.

- `remove(Element)`

Falls das Element innerhalb der Liste vorkommt, wird das erste vorkommende Element innerhalb der Liste gelöscht.

- `append(Element)`

Ein neues Element wird am Ende der Liste hinzugefügt.

- `index(Element)` <a class="anchor" id="index-listen"></a>

Die Position des Vorkommens eines bestimmten Elements (hier: `Paris`) wird mit der Funktion `index()` ermittelt. So erhalten wir die Nummerierung des Elements `Paris`.
**Achtung**: Wenn das Element mehrmals in der Liste vorkommt, gibt uns die Funktion `index()` immer die **Nummerierung** vom **ersten Element** zurück.

## 4.4 Wahrheitswerte <a class="anchor" id="booleans"></a>

Elemente und Ausdrücke (bsp. `3 + 5 > 6`) können wahr oder falsch sein. Diese Ausdrücke liefern eines der beiden Schlüsselwörter `True` (wahr) oder `False` (falsch). Dies sind die einzigen Objekte des Datentyps `bool`.

In [72]:
# True und False
W = True
print("Wahrheitswert:", W)
W = False
print("Wahrheitswert:", W)

Wahrheitswert: True
Wahrheitswert: False


### Vergleichsoperatoren <a class="anchor" id="vergleichsoperatoren"></a>

Besonders im Zusammenhang mit Bedingungsprüfungen (`if`, `while`) wird der Wahrheitswert eines Ausdrucks benötigt.

Bespiel: Falls eine Zahl grösser als 10 ist, sollen bestimmte Anweisungen ausgeführt werden. Der dabei benötigte Ausdruck `x > 10` ist wahr, wenn `x` einen Zahlenwert grösser als 10 hat. Er ist falsch, wenn `x` einen Zahlenwert kleiner oder gleich 10 hat.

Hier haben wir den Vergleichsoperator `>` verwendet. Folgende Vergleichoperatoren sind wichtig:

- `>` : strikt grösser
- `>=`: grösser oder gleich
- `<` : strikt kleiner 
- `<=`: kleiner oder gleich
- `==`: gleich
- `!=`: ungleich

In [73]:
# strikt grosser/kleiner
W = 5>3
print("5>3:", W)
W = 5<3
print("5<3:", W)

# grösser/kleiner oder gleich
W = 5 >= 5
print("5 >= 5:", W)
W = 5 <= 4
print("5 <= 6:", W)

# gleich /ungleich
W = 'Hallo' == 'Hallo'
print("Hallo == Hallo:", W)
W = "Hallo" != 'Welt'
print("Hallo != Welt:", W)

# Datentyp
W = 5>3
print("Typ von 5>3: ", type(W))

5>3: True
5<3: False
5 >= 5: True
5 <= 6: False
Hallo == Hallo: True
Hallo != Welt: True
Typ von 5>3:  <class 'bool'>


### Wahrheitswerte von Objekten

Alle Objekte in Python besitzen einen Wahrheitswert. Der Wahrheitswert eines Objektes kann mit der Funktion `bool()` überprüft werden (wie die Funktion `type()`).

Folgende Objekte sind wahr (liefern `True`):

- eine Zahl ungleich 0 (grösser oder kleiner als 0)
- eine nicht leere Zeichnkette
- eine nicht leere Liste

Folgende Objekte sind falsch (liefern `False`):

- eine Zahl gleich 0
- eine leere Zeichnkette
- eine leere Liste

Beispiel:

In [74]:
# wahre Zahl
Z = 5 + 0.001 - 5
print("Zahl", Z, "ist", bool(Z))

# nicht wahre Zahl
Z = 5.75 - 5.75
print("Zahl", Z, "ist", bool(Z))

# String
S = "Kurt"
print("String", S, "ist nicht leer, also", bool(S))

# Wahre Liste
L = [3,4]
print("Liste", L, "ist nicht leer, also", bool(L))

# Leere Liste
print("Liste", [], "ist leer also", bool([]))

Zahl 0.001000000000000334 ist True
Zahl 0.0 ist False
String Kurt ist nicht leer, also True
Liste [3, 4] ist nicht leer, also True
Liste [] ist leer also False


# 5. Programmierkurs <a class="anchor" id="programmierkurs"></a>

In diesem Abschnitt wollen wir noch einmal grundlegende Konzepte der Programmierung wiederholen. Wir konzentrieren uns auf einfache Verzweigungen und Schleifen.

## 5.1 Verzweigungen <a class="anchor" id="verzweigungen"></a>

In den bisherigen Code-Zellen werden alle Anweisungen der Reihe nach ausgeführt. Zur Steuerung des Programmablaufs werden allerdings häufig Verzweigungen benötigt. 
Innerhalb des Programms wird anhand einer Bedingung entschieden, welcher Zweig des Programms ausgeführt wird.<br>

Wenn du beispielsweise das Menu in der Mensa nur dann zum Mittagessen nimmst, wenn weniger als 10 Leute anstehen, würden wir sagen:

- *Falls weniger als 10 Leute anstehen*:
    - *Essen in der Mensa*
- *Andernfalls (sonst)*
    - *Essen ausserhalb der Kanti*

Genau gleich funktionieren Verzweigungen in der Programmierung.

In diesem Beispiel ist *weniger als 10 Leute anstehen* die **Bedingung** und *Essen in der Mensa* resp. *Essen ausserhalb der Kanti* die jeweiligen **Zweige** die ausgeführt werden, falls die Bedingung `wahr` oder `falsch` ist.

Bedingungen werden mithilfe von [**Vergleichsoperatoren**](#vergleichsoperatoren) durchgeführt, und resultieren in einem [Wahrheitswert](#booleans). <br>
Wenn die Bedingung wahr ist, wird der Zweig der `if` Bedingung durchgeführt, ansonsten wird der Zweig innerhalb von `else` durchgeführt.

*Syntax*:

```python
if BEDINGUNG:
    # Zweig wenn Bedingung wahr ist
else:
    # Zweig wenn Bedingung falsch ist
```

*Wichtig*:<br>
Nach der Bedingung kommt immer ein Doppelpunkt `:`. Die Anweisungen müssen innerhalb des sogenannten `if`-Zweigs mithilfe der `Tab`-Taste eingerückt werden, damit Python die Zugehörigkeit zur Verzweigung erkennen kann. Gleichzeitig macht die Einrückung das Programm übersichtlicher.

Erstes Beispiel:

In [75]:
x = 12
print("x:", x)

# Die Bedingung
if x > 0:
    print("Diese Zahl ist positiv")
else:
    print("Diese Zahl ist 0 oder negativ")

x: 12
Diese Zahl ist positiv


Ein `if`-Zweig kann auch ohne `else`-Zweig existieren. <br>
Das ist der Fall, wenn wir aufgrund einer Bedingung etwas ausführen wollen. Wenn die Bedingung jedoch falsch ist, wird einfach nichts ausgeführt.

Intuitives Beispiel:

- *Falls es heute nicht regnet*:
    - *Das Joggingtraining findet statt*

Wenn die Bedingung im `if`-Statement erfüllt ist, wird der `if`-Zweig durchgeführt, ansonsten nicht.

In [76]:
x = 18
print("x:", x)

# Die Bedingung wird erfüllt wenn x % 3 == 0 ist. Das heisst, wenn x / 3 eine Division ohne Rest ist.
if (x % 3) == 0:
    print("3 ist ein Teiler der Zahl", x)

x = 20

# Die Bedingung wird nicht mehr erfüllt sein, da 20 / 3 = 6 R 2, 
# also 20 % 3 = 2, und deshalb ist die Bedingung falsch
if (x % 3) == 0:
    print("3 ist ein Teiler der Zahl", x)

x: 18
3 ist ein Teiler der Zahl 18


### Mehrfache Verzweigung

In vielen Anwendungsfällen gibt es mehr als zwei Möglichkeiten, zwischen denen zu entscheiden ist. Dazu wird eine mehrfache Verzweigung benötigt. Im folgenden Beispiel wird untersucht, ob eine Zahl positiv, negativ oder gleich 0 ist. Es wird eine entsprechende Meldung ausgegeben.

In [77]:
x = -5
print("x:", x)

if x > 0:
    print("x ist positiv")
elif x < 0:
    print("x ist negativ")
else:
    print("x ist gleich 0")

x: -5
x ist negativ


- `elif`

Nach `elif` wird eine weitere Bedingung formuliert. Sie wird nur untersucht, falls die erste Bedingung (nach dem `if`) nicht zutrifft. Falls `x` negativ ist, werden die nachfolgenden, eingerückten Anweisungen ausgeführt.

- `else`

Die Anweisungen nach dem `else` werden nur durchgeführt, falls keine der beiden vorherigen Bedingungen zutraf.

## 5.2 Schleifen <a class="anchor" id="schleifen"></a>

Neben der Verzweigung gibt es eine weitere wichtige Struktur zu Steuerung von Programmen: die Schleife. Mithilfe einer Schleife ermöglichst du die wiederholte Ausführung eines Programmschritts.

Es muss zwischen zwei Typen von Schleifen unterschieden werden: der `for`-Schleife und der `while`-Schleife. Der jeweilige Anwendungsbereich der beiden Typen wird durch folgende Merkmale definiert:

- `for`

Eine `for`-Schleife wird verwendet, wenn ein Programmschritt für eine regelmässige, zum Zeitpunkt der Anwendung bekannte Folge von Werten wiederholt ausgeführt werden soll. Beispielweise kannst du eine `for`-Schleife verwenden, wenn du für alle Elemente innerhalb einer Liste eine Berechnung durchführen willst. Oder wenn du für alle Zahlen zwischen 1 und 10, eine Berechnung mit der jeweiligen Zahl durchführen willst. Deshalb nennen wir die `for`-Schleife auch die *Zählschleife*.

- `while`

Eine `while`-Schleife ist eine sogenannte *bedingungsgesteuerte Schleife*. Sie wird verwendet, wenn du ein Programmschritt aufgrund von einer Bedinung mehrmahls durchführen willst. Sobald die Bedingung nicht mehr erfüllt ist, wird der Programmschritt nicht mehr durchgeführt.

### for-Schleife <a class="anchor" id="for-schleife"></a>

Betrachte das folgende Beispiel:

In [78]:
x = 1
print("Das Quadrat von", x, "ist", x*x)
x = 2
print("Das Quadrat von", x, "ist", x*x)
x = 3
print("Das Quadrat von", x, "ist", x*x)
x = 4
print("Das Quadrat von", x, "ist", x*x)
x = 5
print("Das Quadrat von", x, "ist", x*x)

Das Quadrat von 1 ist 1
Das Quadrat von 2 ist 4
Das Quadrat von 3 ist 9
Das Quadrat von 4 ist 16
Das Quadrat von 5 ist 25


Wie du vielleicht erkennst, verändern wir den Wert von `x`. Mit jedem neuen Wert wird eine Ausgabe mit dem Quadrat des Wertes generiert.

Eine `for`-Schleife kann diese Code-Zelle aber vereinfachen. Beispielsweise könnten wir sagen:

- **für** *Zahl* `in` *alle Zahlen zwischen 1 und 5*:
    - print("Das Quadrat von", Zahl, "ist", Zahl*Zahl)

Also, wiederhole diesen Programmschritt für alle Zahlen im Interval [1, 5].

Die `for`-Schleife führt dann diesen Programmschritt 5-Mal durch. Damit die `for`-Schleife aber bei einer **spezifischen Wiederholung** weiss, welche Zahl momentan gemeint ist, müssen wir eine Variable definieren, die innerhalb der `for`-Schleife den Wert der aktuellen Zahl hält.

**Syntax**:

```python

for variable in Sequenz:
    # berechne etwas mit der Variable x
```

Du erkennst hier, dass ich bei der Syntax bewusst das Wort **Sequenz** geschrieben habe. Denn bei `for`-Schleifen geben wir das Intervall oft als **Sequenz** an (Liste).

### Range

Eine wichtige **Sequenz** im Zusammenhang mit `for`-Schleifen erzeugen wir mit der sogenannten `range(startwert, endwert)` Funktion. Sie ermöglicht uns eine **Liste von ganzen Zahlen** zu erstellen. Dabei gilt:

**Startwert**: Erste Zahl die zur Liste (Intervall) gehört<br>
**Endwert**: Erster Zahl die **NICHT** mehr zur Liste (Intervall) gehört.

Wenn du die Funktion `range` nur mit **einem Wert** verwendest, wird dieser Wert als **Endwert** angeschaut. 
Das heisst, `range(9)` beinhaltet alle Zahlen von 0 - 8. (Auch hier starten wir immer bei 0!)

So können wir das obige Programm folgendermassen realisieren:

In [79]:
# Das Interval von [1, 5] erreichen wir mit range(1, 6)
# Der Name der Variable ist i
for i in range(1, 6):
    print("Das Quadrat von", i, "ist", i*i)

Das Quadrat von 1 ist 1
Das Quadrat von 2 ist 4
Das Quadrat von 3 ist 9
Das Quadrat von 4 ist 16
Das Quadrat von 5 ist 25


Ich kann für die Variable natürlich auch einen anderen Namen wählen.

In [80]:
# Das Interval von [1, 5] erreichen wir mit range(1, 6)
# Der Name der Variable ist zahl
for zahl in range(1, 6):
    print("Das Quadrat von", zahl, "ist", zahl*zahl)

Das Quadrat von 1 ist 1
Das Quadrat von 2 ist 4
Das Quadrat von 3 ist 9
Das Quadrat von 4 ist 16
Das Quadrat von 5 ist 25


Diese `for`-Schleife verläuft so ab:

- Nimm den Wert des **ersten** Elements innerhalb der Sequenz
    - Speichere den Wert in der Variable `zahl` (Das erste Element ist die Zahl `0`)
    - Führe alle Berechnungen innerhalb der Schleife durch
- Nimm den Wert des **zweiten** Elements innerhalb der Sequenz
    - Speichere den Wert in der Variable `zahl` (Das erste Element ist die Zahl `1`)
    - Führe alle Berechnungen innerhalb der Schleife durch
- Nimm den Wert des **dritten** Elements innerhalb der Sequenz
    - Speichere den Wert in der Variable `zahl` (Das erste Element ist die Zahl `2`)
    - Führe alle Berechnungen innerhalb der Schleife durch
- etc.

Dieses Verfahren wird für alle Zahlen wiederholt die im Intervall [1, 5] liegen (angegeben mit `range(1, 6)`)

### Break und Continue

Wir können Schleifen im allgemeinen unterbrechen, falls eine gewisse Beidnung erfüllt ist (`break`). 
Zudem können wir aufgrund von einer Bedingung auch sagen, dass für **das aktuelle Element** der Programmschritt **nicht durchgeführt** werden muss, ohne die Schleife abzubrechen (`continue`).

Beispiele:

In [81]:
# Das Interval von [1, 5] erreichen wir mit range(1, 6)
# Der Name der Variable ist zahl
for zahl in range(1, 6):
    # Die for-Schleife bricht ab, sobald das Quadrat grösser ist als 15
    if zahl * zahl > 15:
        # sobald eine Zahl gefunden wurde, die im Quadrat grösser ist als 15, wird die for-Schleife abgebrochen und die Berechnung wird für die übrigen Zahlen nicht mehr durchgeführt
        break
    print("Das Quadrat von", zahl, "ist", zahl*zahl)

Das Quadrat von 1 ist 1
Das Quadrat von 2 ist 4
Das Quadrat von 3 ist 9


In [82]:
# Das Interval von [1, 5] erreichen wir mit range(1, 6)
# Der Name der Variable ist zahl
for zahl in range(1, 6):
    # Solange das Quadrat der Zahl kleiner ist als 15, fahren wir mit der for-Schleife direkt weiter,
    # das heisst, 
    # alle darauffolgenden Zeilen innerhalb der for-Schleife, werden für diesen Durchlauf nicht mehr ausgeführt
    if zahl * zahl < 15:
        # Fahre in der Schleife direkt mit der nächsten Zahl weiter
        continue
    # Dieses Statement wird nur ausgeführt wenn das Quadrat der Zahl grösser oder gleich 15 ist
    print("Das Quadrat von", zahl, "ist", zahl*zahl)

Das Quadrat von 4 ist 16
Das Quadrat von 5 ist 25


### `for`-Schleifen mit Listen

Wir können nun ganz einfach über Listen iterieren. Wie wir bereits wissen, geben wir das Interval der `for`-Schleife mit einer Sequenz an.

Dazu gibt es zwei Möglichkeiten:

- Wir iterieren über die **Nummerieung** aller Elemente in der Liste
    - In diesem Fall müssen wir noch den Wert des Elements mit der jeweiligen Nummerierung innerhalb der `for`-Schleife abfragen

In [83]:
staedte = ["Bern", "Basel", "Olten", "Zürich"]

print("Beispiel 1")

# für die Zahl 0 - (Anzahl elemente in staedte - 1)
for nummerierung in range(len(staedte)):
    # Da die Variable nummerierung eigentlich nur den aktuellen Index enthält, 
    # müssen wir den Wert innerhalb der Liste noch mit staedte[nummerierung] abfragen
    print("Stadt:", staedte[nummerierung])

Beispiel 1
Stadt: Bern
Stadt: Basel
Stadt: Olten
Stadt: Zürich


- Wir iterieren direkt über die **Werte** aller Elemente in der Liste
    - In diesem Fall wissen wir nicht genau welche **Nummerierung** das Element innerhalb der Liste hat, wir haben aber bereits den Wert

In [84]:
print("Beispiel 2")
# Direkter Weg: Hier enthält der Wert der Variable stadt direkt den Wert des Elements innerhalb der Liste
for stadt in staedte:
    print("Stadt:", stadt)

Beispiel 2
Stadt: Bern
Stadt: Basel
Stadt: Olten
Stadt: Zürich


Überlege dir wie du in dieser `for`-Schleife die Nummerierung vom Element innerhalb der Liste abfragen kannst.
[Tipp (Index)](#index-listen)

Stelle sicher, dass du die Unterschiede der beiden Beispiele gut verstehst.

### Enumerate

Alternativ gibt es noch eine Möglichkeit, wie wir über die Elemente innerhalb der Liste **und** deren Nummerierung zusammen itertieren können.
Dazu existiert die eingebaute Funktion `enumerate(Liste)`.

**Beispiel:**

```python
    for nummerierung, stadt in enumerate(staedte):
        print("Stadt", stadt, "mit Nummerierung", nummerierung)
```

**Wichtig**: Die erste Variable bezieht sich in jedem Fall auf die **Nummerierung** und die zweite Variable auf das **Element** innerhalb der Liste.

In [85]:
# Einfacher weg, wenn wir die Nummerierung des Elements in der Liste und den Wert der Liste innerhalb der gleichen Schleife haben wollen
for nummerierung, stadt in enumerate(staedte):
    print("Stadt", stadt, "mit Nummerierung", nummerierung)

Stadt Bern mit Nummerierung 0
Stadt Basel mit Nummerierung 1
Stadt Olten mit Nummerierung 2
Stadt Zürich mit Nummerierung 3


### while-Schleife <a class="anchor" id="while-schleife"></a>

Die `while`-Schleife dient zur Steuerung einer Wiederholung mithilfe einer Bedingung. Im folgenden Programm werden zufällige Zahlen addiert und ausgegeben. Solange die Summe der Zahlen kleiner als 30 ist, wird der Vorgang wiederholt. Ist die Summe gleich oder grösser als 30, wird das Programm beendet.

Syntax while-Schleife:

```python

while BEDINGUNG:
    # While-Block
```

Beispiel:

In [91]:
# Zufallsgenerator
import random
random.seed()

# Initialisierung
summe = 0

# while-Schleife
while summe < 30:
    zahl = random.randint(1,8)
    summe = summe + zahl
    print("Zahl:", zahl, "Zwischensumme:", summe)
    
print("Ende")

Zahl: 6 Zwischensumme: 6
Zahl: 8 Zwischensumme: 14
Zahl: 2 Zwischensumme: 16
Zahl: 8 Zwischensumme: 24
Zahl: 3 Zwischensumme: 27
Zahl: 2 Zwischensumme: 29
Zahl: 7 Zwischensumme: 36
Ende


### Liste iterieren mit while

Wir können auch `while`-Schleifen verwenden um über Listen zu iterieren.
Dazu musst du jeweils mit der Nummerierung innerhalb einer Liste arbeiten.<br>

Du brauchst eine Zählvariable, mit der du immer wieder überprüfst, ob die Liste die Nummerierung der aktuellen Zählvariable beinhaltet. Die Zählvariable muss aber in jedem Schritt erhöht werden.

Beispiel:

In [92]:
staedte = ["Bern", "Zürich", "Olten", "Solothurn", "Genf", "Lausanne"]

# starte mit einer Zählvariable
i = 0
# währendem die Zählvariable kleiner ist als die Länge der Liste 'staedte'
while i < len(staedte):
    print("Die Stadt mit Nummerierung", i, "ist", staedte[i])
    i += 1

Die Stadt mit Nummerierung 0 ist Bern
Die Stadt mit Nummerierung 1 ist Zürich
Die Stadt mit Nummerierung 2 ist Olten
Die Stadt mit Nummerierung 3 ist Solothurn
Die Stadt mit Nummerierung 4 ist Genf
Die Stadt mit Nummerierung 5 ist Lausanne


# 6. Weiterführende Programmierung <a class="anchor" id="weiterfuehrende-programmierung"></a>

## 6.1 Funktionen <a class="anchor" id="funktionen"></a>

Die Modularisierung, also die Zerlegung einer Programms in selbst geschriebene Funktionen, bietet besonders bei grösseren Programmen unübersehbare Vorteile:

- Programmteile die mehrmals benötigt werden, müssen nur einmal definiert werden
- Nützliche Programmteile können in mehreren Programmen verwendet werden.
- Umfangreiche Programme können in unterschiedliche Teile zerlegt werden.
- Pflege und Wartung von Programmen wird erleichtert.
- Der Programmcode ist für die Programmiererin/den Programmierer leichter zu verstehen.

### Funktionsdefinition und Funktionsaufruf

Funktionen können uns also helfen, den Programmcode zu vereinfachen. Dabei müssen wir unterscheiden:

- **Funktionsdefinition**:
    Hier definieren wir die Aktion, welche von dieser Funktion ausgeführt werden soll. Dabei wird die Aktion bei der Funktionsdefinition selber nicht ausgeführt, sondern nur *definiert*. Wenn wir die Aktion der Funktion ausführen wollen, brauchen wir einen **Funktionsaufruf**.<br>
    <br>
- **Funktionsaufruf**:
    Wenn die Funktion verwendet wird (wie bei `print()`), wird automatisch der Programmcode innerhalb der **Funktionsdefinition** ausgeführt.
<br>
<br>
**Wichtig**: <br>
Eine Funktion kann erst dann verwendet werden (*Funktionsaufruf*), wenn Sie zuvor definiert wurde.

#### Schreibweise 

Schreibweise für die Funktionsdefinition:

```python
    def funktionsname():
        # Funktionsblock
```

Eine Funktion wird in Python mit dem Schlüsselwort **def** angezeigt. <br>
Danach geben wir der Funktion einen Namen (diesen Namen verwenden wir auch beim Aufruf).<br>
Nach dem Namen folgen einfache Klammern `()` (wie du später sehen wirst, ist das ein Behälter für Parameter) und ein Doppelpunkt `:`.<br>
Nun gehört **alles** innerhalb vom **Funktionsblock** zur ausführbaren Aktion der Funktion. Das heisst, beim Funktionsaufruf wird der ganze **Funktionsblock** ausgeführt.<br>

**Wichtig**: Alle Zeilen innerhalb vom **Funktionsblock** müssen mit der *Tab*-Taste eingerückt werden.

Schreibweise für die Funktionsaufruf:

Wenn wir die Funktion einmal definiert haben, können wir sie aufrufen. Wenn wir die Funktion aufrufen, wird der gesamte **Funktionsblock** der Funktionsdefinition ausgeführt.

**Wichtig**: Beim Funktionsaufruf musst du ebenfall **immer** die einfachen Klammern `()` verwenden, damit der Python interpreter weiss, dass du eine Funktion ausführen willst.

```python
    funktionsname()
```

### Einfache Funktionen

Einfache Funktionen führen beim **Aufruf** stets die gleiche Aktion aus. Im folgenden Beispiel führt jeder Aufruf der Funktion `stern()` dazu, dass eine optische Trennung auf dem Bildschirm ausgegeben wird. 

Ein Beispiel:

In [86]:
# Definition der Funktion
def stern():
    print("-----------------")
    print("*** Trennung ****")
    print("-----------------")

# Haupt-Programm
x = 12
y = 5
stern()                         # 1. Aufruf
print("x =", x, ", y =", y)
stern()                         # 2. Aufruf
print("x + y =", x + y)
stern()                         # 3. Aufruf
print("x - y =", x - y)
stern()                         # 4. Aufruf

-----------------
*** Trennung ****
-----------------
x = 12 , y = 5
-----------------
*** Trennung ****
-----------------
x + y = 17
-----------------
*** Trennung ****
-----------------
x - y = 7
-----------------
*** Trennung ****
-----------------


Die Funktion `stern()` wird zuerst nur *definiert* und nicht *ausgeführt*. Sie steht sozusagen zum späteren Gebrauch bereit. Im unteren Teil der Zelle beginnt das eigentliche Programm.

*Übrigens*: Den Namen einer Funktion kannst du weitgeend frei wählen.

### Parameter <a class="anchor" id="parameter"></a>

Einfache Funktionen können wir daran erkennen, dass sie von keiner *äusseren* Variable abhängig sind. Das heisst, sie führen beim Aufruf stets die gleiche *Aktion* aus. Oft ergibt es jedoch Sinn, wenn wir der Funktion auch noch **Werte** übergeben können, so dass die *Aktion* der Funktion abhängig von diesen übergebenen **Werten** ist. 

Diese *Informationen / Werte* die wir an die Funktion übermitteln nennen wir **Parameter**. 

Dazu ein kleines Beispiel:

In [87]:
ersteZahl = 18

if (ersteZahl % 3) == 0:
    print("3 ist ein Teiler der Zahl", ersteZahl)
if (ersteZahl % 5) == 0:
    print("5 ist ein Teiler der Zahl", ersteZahl)
if (ersteZahl % 2) == 0:
    print("2 ist ein Teiler der Zahl", ersteZahl)

zweiteZahl = 20

if (zweiteZahl % 3) == 0:
    print("3 ist ein Teiler der Zahl", zweiteZahl)
if (zweiteZahl % 5) == 0:
    print("5 ist ein Teiler der Zahl", zweiteZahl)
if (zweiteZahl % 2) == 0:
    print("2 ist ein Teiler der Zahl", zweiteZahl)

3 ist ein Teiler der Zahl 18
2 ist ein Teiler der Zahl 18
5 ist ein Teiler der Zahl 20
2 ist ein Teiler der Zahl 20


Wie du sehen kannst, definiere ich zwei Variablen `ersteZahl` und `zweiteZahl`. Für beide Zahlen will ich gewisse Teiler ermitteln. Deshalb muss ich 6 Programmzeilen 2-Mal schreiben, die sich eigentlich nur in der verwendeten Variable (`ersteZahl` und `zweiteZahl`) innerhalb der Bedingung und der `print()` Funktion unterscheiden.

Nun können wir für diese 6-zeilen eine Funktion definieren, die beim Aufruf genau diese 6-Zeilen ausführt. 
Diese 6-Zeilen sind jedoch abhängig vom Wert einer Variable. Da ich bei der Funktionsdefinition diese Variable nicht kenne, muss ich Sie als **Parameter** in der Funktionsddefinition kennzeichnen. Die **Parameter** einer Funktion werden in die einfachen Klammern `()` geschrieben. 

Du kannst den Namen der Parameter frei wählen.

**Syntax**:

```python

def Teiler(parameter):
    # Funktionsblock
```

Für diese 6-Zeilen könnten wir somit folgende Funktion definieren:

In [88]:
# Funktionsdefinition mit einem Parameter
def Teiler(untersuchendeZahl):
    print("Untersuche Teiler der Zahl", untersuchendeZahl)
    if (untersuchendeZahl % 3) == 0:
        print("3 ist ein Teiler der Zahl", untersuchendeZahl)
    if (untersuchendeZahl % 5) == 0:
        print("5 ist ein Teiler der Zahl", untersuchendeZahl)
    if (untersuchendeZahl % 2) == 0:
        print("2 ist ein Teiler der Zahl", untersuchendeZahl)

# Variablendefinition
ersteZahl = 18
zweiteZahl = 20
dritteZahl = 21

# Funktionsaufruf
Teiler(ersteZahl)
Teiler(zweiteZahl)
Teiler(dritteZahl)

Untersuche Teiler der Zahl 18
3 ist ein Teiler der Zahl 18
2 ist ein Teiler der Zahl 18
Untersuche Teiler der Zahl 20
5 ist ein Teiler der Zahl 20
2 ist ein Teiler der Zahl 20
Untersuche Teiler der Zahl 21
3 ist ein Teiler der Zahl 21


Wie du sehen kannst, geben wir dem Parameter in der Funktionsdefinition den Namen `untersuchendeZahl`. Nun sind alle Berechnungen innerhalb des Funktionsblocks abhängig von diesem einen **Parameter**.

Nach der Definition können wir die Funktion ausführen und erhalten je nach übergebenem Wert eine andere Ausgabe. Die Aktion der Funktion ist also abhängig von der übergebenen Information.

#### Funktion mit mehreren Parameter

Ein Funktion kann auch noch vielseitiger werden. Wir können beispielsweise mehrere Parameter an die Funktion übermitteln. Das bedeutet aber auch, dass die Funktionsdefinition mehrere Parameter beinhalten muss.

Beispielweise könnte ich die obige Funktion noch Vereinfachen, um Teiler zu untersuchen.

In [89]:
# Funktionsdefinition mit mehreren Parametern
def Teiler_neu(dividend, divisor):
    if (dividend % divisor) == 0:
        print(divisor, "ist ein Teiler der Zahl", dividend)
    else:
        print(divisor, "ist kein Teiler der Zahl", dividend)

# Funktionsaufruf
Teiler_neu(18, 2)
Teiler_neu(18, 4)
Teiler_neu(18, 5)
Teiler_neu(18, 6)

2 ist ein Teiler der Zahl 18
4 ist kein Teiler der Zahl 18
5 ist kein Teiler der Zahl 18
6 ist ein Teiler der Zahl 18


### Rückgabewerte <a class="anchor" id="rueckgabewerte"></a>

Funktionen werden häufig zur Berechnung von Ergebnissen eingesetzt. Zu diesem Zweck können Funktionen ihre Ergebnisse als sogenannte *Rückgabewerte* zurückliefern.<br>

Diese Eigenschaft kann nützlich sein, wenn wir beispielsweise das Ergebnis eines Funktionaufrus in einer neuen Variable speichern wollen. 

Syntax Rückgabewerte:

```python
    def funktionsname():
        # Funktionsblock
        return ergebnis
```

Der Rückgabewert wird mit dem **return** Schlüsselwort gekennzeichnet. 

Beispiel:

In [90]:
# Definition der Funktion
def mittelwert(x,y):
    ergebnis = (x+y) / 2
    return ergebnis

# Hauptprogramm
c = mittelwert(3, 9)
print("Mittelwert:", c)

Mittelwert: 6.0


Innerhalb der Funktion wird zunächst das Ergebnis berechnet. Es wird anschliessend mithilfe der Anweisung *return* an die aufrufende Stelle zurückgeliefert. Die Anweisung *return* beendet ausserdem unmittelbar den Ablauf der Funktion.

**Rückgabe speichern**:

Beim ersten Aufruf wird der Rückgabewert in der Variablen *c* zwischengespeichert. Er kann im weiteren Verlauf des Programms an beliebiger Stelle verwendet werden.

## 6.2 Eingebaute Funktionen <a class="anchor" id="eingebaute-funktionen"></a>

## 6.3 Ausgabe und Formatierung <a class="anchor" id="ausgabe-formatierung"></a>

TODO



### Funktion print() <a class="anchor" id="print"></a>

TODO



## 6.4 Module <a class="anchor" id="module"></a>

TODO